src/storage/src/hummock/sstable/xor_filter.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use bytes::{Buf, BufMut};
+use itertools::Itertools;
+use xorf::{Filter, Xor16};
+
+use super::{FilterBuilder, Sstable};
+
+pub struct XorFilterBuilder {
+    key_hash_entries: Vec<u64>,
+}
+
+impl XorFilterBuilder {
+    pub fn new(capacity: usize) -> Self {
+        let key_hash_entries = if capacity > 0 {
+            Vec::with_capacity(capacity)
+        } else {
+            vec![]
+        };
+        Self { key_hash_entries }
+    }
+}
+
+impl FilterBuilder for XorFilterBuilder {
+    fn add_key(&mut self, key: &[u8], table_id: u32) {
+        self.key_hash_entries
+            .push(Sstable::hash_for_bloom_filter(key, table_id));
+    }
+
+    fn approximate_len(&self) -> usize {
+        self.key_hash_entries.len() * 5
+    }
+
+    fn finish(&mut self) -> Vec<u8> {
+        let xor_filter = Xor16::from(
+            &HashSet::<u64>::from_iter(std::mem::take(&mut self.key_hash_entries).into_iter())
+                .into_iter()
+                .collect_vec(),
```

Comment:
We can use `HashSet` directly for `self.key_hash_entries` to avoid conversion between Vec and HashSet back and forth here.

Reply:
> We can use `HashSet` directly for `self.key_hash_entries` to avoid conversion between Vec and HashSet back and forth here.

There are two reasons:
- Inserting into `HashSet` is slow
- We cannot know exact size in advance (so we cannot call `reserve` with a proper paramater)

Reply:
Got it. Why do we need to convert to `HashSet` in the first place? Xor filter cannot handle duplicates?

Reply:
> Got it. Why do we need to convert to `HashSet` in the first place? Xor filter cannot handle duplicates?

yes, Because Xor filter cannot handle duplicates.